datasets/flwr_datasets/partitioner/shard_partitioner.py

```diff
@@ -0,0 +1,322 @@
+
+        Returns
+        -------
+        dataset_partition : Dataset
+            single partition of a dataset
+        """
+        # The partitioning is done lazily - only when the first partition is
+        # requested. Only the first call creates the indices assignments for all the
+        # partition indices.
+        self._check_num_partitions_correctness_if_needed()
+        self._check_possibility_of_partitions_creation()
+        self._sort_dataset_if_needed()
+        self._determine_node_id_to_indices_if_needed()
+        return self.dataset.select(self._node_id_to_indices[node_id])
+
+    def _determine_node_id_to_indices_if_needed(self) -> None:  # pylint: disable=R0914
+        """Assign sample indices to each node id.
+
+        This method works on sorted datasets. A "shard" is a part of the dataset of
+        consecutive samples (if self._keep_incomplete_shard is False, each shard is same
+        size).
+        """
+        # No need to do anything if that node_id_to_indices are already determined
+        if self._node_id_to_indices_determined:
+            return
+
+        # One of the specification allows to skip the `num_shards_per_node` param
+        if self._num_shards_per_node is not None:
+            self._num_shards_used = int(
+                self._num_partitions * self._num_shards_per_node
+            )
+            num_shards_per_node_array = (
+                np.ones(self._num_partitions) * self._num_shards_per_node
+            )
+            if self._shard_size is None:
+                self._compute_shard_size_if_missing()
+                assert self._shard_size is not None
+                if self._keep_incomplete_shard:
+                    num_usable_shards_in_dataset = int(
+                        math.ceil(len(self.dataset) / self._shard_size)
+                    )
+                else:
+                    num_usable_shards_in_dataset = int(
+                        math.floor(len(self.dataset) / self._shard_size)
+                    )
+            else:
+                num_usable_shards_in_dataset = int(
+                    math.floor(len(self.dataset) / self._shard_size)
+                )
+        elif self._num_shards_per_node is None:
+            if self._shard_size is None:
+                raise ValueError(
+                    "The shard_size needs to be specified if the "
+                    "num_shards_per_node is None"
+                )
+            if self._keep_incomplete_shard is False:
+                self._num_shards_used = int(
+                    math.floor(len(self.dataset) / self._shard_size)
+                )
+                num_usable_shards_in_dataset = self._num_shards_used
+            elif self._keep_incomplete_shard is True:
+                self._num_shards_used = int(
+                    math.ceil(len(self.dataset) / self._shard_size)
+                )
+                num_usable_shards_in_dataset = self._num_shards_used
```

Comment:
Here `self._num_shards_used` could be < `self._num_partitions` if passes a big `shard_size`. The code runs, but some partitions will be empty. What's the best way to proceed? throw warning and leave partitions empty? or throw exception and ask for lower `shard_size` so at least each partition gets 1 shard?

Reply:
Thanks for catching that. It needs to be covered. 
I think the best option is to throw an exception because if not, then some strange things might happen when someone tries to train on an empty dataset. 

I'll also extend the test to capture that.

Reply:
Done